jbmc/src/java_bytecode/select_pointer_type.cpp

```diff
@@ -109,3 +127,98 @@ pointer_typet select_pointer_typet::specialize_generics(
+/// \param generic_specialization_map Map of type names to specialization stack
+/// \return The first instantiated type for the generic type or nothing if no
+/// such instantiation exists.
+optionalt<pointer_typet>
+select_pointer_typet::get_recursively_instantiated_type(
+  const irep_idt &parameter_name,
+  const generic_parameter_specialization_mapt
+    &generic_parameter_specialization_map) const
+{
+  generic_parameter_recursion_trackingt visited;
+  size_t depth = 0;
+  const size_t max =
+    generic_parameter_specialization_map.find(parameter_name)->second.size();
+
+  irep_idt current_parameter = parameter_name;
+  while(depth < max)
+  {
+    const auto retval = get_recursively_instantiated_type(
+      current_parameter, generic_parameter_specialization_map, visited, depth);
+    if(retval.has_value())
+    {
+      CHECK_RETURN(!is_java_generic_parameter(*retval));
+      return retval;
+    }
+    CHECK_RETURN(visited.empty());
+
+    const auto &entry =
+      generic_parameter_specialization_map.find(current_parameter)
+        ->second.back();
+    const java_generic_parametert &gen_param = to_java_generic_parameter(entry);
+    const auto &type_var = gen_param.type_variable();
+    current_parameter = type_var.get_identifier();
+    depth++;
+  }
+  return {};
+}
+
+/// See get_recursively instantiated_type, the additional parameters just track
+/// the recursion to prevent, visiting the same depth again and specify which
+/// stack depth is analyzed.
+/// \param visited Tracks the visited parameter names
+/// \param depth Stack depth to analyze
+optionalt<pointer_typet>
+select_pointer_typet::get_recursively_instantiated_type(
+  const irep_idt &parameter_name,
+  const generic_parameter_specialization_mapt
+    &generic_parameter_specialization_map,
+  generic_parameter_recursion_trackingt &visited,
+  const size_t depth) const
+{
+  // Get the pointed to instantiation type at the desired stack depth.
+  // - if this type is not a generic type, it is returned as a valid
+  //   instantiation
+  // - if nothing can be found at the given depth, an empty optional is returned
+
+  auto val = generic_parameter_specialization_map.find(parameter_name);
+  INVARIANT(
+    val != generic_parameter_specialization_map.end(),
+    "generic parameter must be a key in map");
+
+  const auto &replacements = val->second;
+
+  INVARIANT(
+    depth < replacements.size(), "cannot access elements outside stack");
+
+  // Check if there is a recursion loop, if yes return with nothing found
+  if(visited.find(parameter_name) != visited.end())
+  {
+    return {};
+  }
+
+  const size_t index = (replacements.size() - depth) - 1;
+  const auto &type = replacements[index];
+
+  if(!is_java_generic_parameter(type))
+  {
+    return type;
+  }
+
+  visited.insert(parameter_name);
+  const auto &gen_type = to_java_generic_parameter(type).type_variable();
+  const auto inst_val = get_recursively_instantiated_type(
+    gen_type.get_identifier(),
```

Comment:
You can use `to_java_generic_parameter(type).get_name()` to get the parameter name directly, no need for an extra variable.